src/sui/index.ts

```diff
@@ -0,0 +1,280 @@
+    const shareBalance = await this.getShareBalance(ownerAddress, shareType);
+    return Number(shareBalance.totalBalance) / Number(oneShare);
+  }
+
+  // returns human readable numeric value for 1 share of the vault
+  async fetchShareValue(accountantId: string) {
+    const accountant = await this.client.getObject({
+      id: accountantId,
+      options: {
+        showContent: true,
+      },
+    });
+    const fields = (accountant.data?.content as any)?.fields;
+    return Number(fields?.exchange_rate) / Number(fields?.one_share);
+  }
+
+  // returns human readable numeric value for TVL in terms of the base asset of the vault
+  async fetchTotalAssets(accountantId: string) {
+    const accountant = await this.client.getObject({
+      id: accountantId,
+      options: {
+        showContent: true,
+      },
+    });
+    const fields = (accountant.data?.content as any)?.fields;
+    const one_share = Number(fields?.one_share);
+    const total_shares = Number(fields?.total_shares);
+    const share_value = Number(fields?.exchange_rate);
+    // assuming exchage rate is in terms of 1 share
+    return (total_shares * share_value) / (one_share * one_share);
+  }
+
+  async fetchRequestUnlockTime(requestId: string) {
+    const request = await this.client.getObject({
+      id: requestId,
+      options: {
+        showContent: true,
+      },
+    });
+    const fields = (request.data?.content as any)?.fields;
+    return fields?.creation_time_ms + fields?.ms_to_maturity;
+  }
+
+  async isVaultPaused(vaultId: string) {
+    const vault = await this.client.getObject({
+      id: vaultId,
+      options: {
+        showContent: true,
+      },
+    });
+    return (vault.data?.content as any)?.fields?.is_vault_paused;
+  }
+
+  // Helper function to read queue key fields
+  async readQueueKeyFields(queueKeyId: string) {
+    const { data } = await this.client.getObject({
+      id: queueKeyId,
+      options: { showContent: true },
+    });
+
+    // the object is `Field<QueueKey, …>`
+    const nameFields = (data as any).content.fields.name.fields;
+    return {
+      account: nameFields.account as string,
+      timestamp: BigInt(nameFields.timestamp as string),
+    };
+  }
+
+  //== Private helper functions ==
+
+  async #getGenericTypeFromObject(objectId: string) {
+    const objectData = await this.client.getObject({
+      id: objectId,
+      options: {
+        showType: true,
+      },
+    });
+  
+    if (!objectData.data?.type) {
+      throw new Error('Type information not found.');
+    }
+  
+    const typeString = objectData.data.type;
+    console.log('Full type string:', typeString);
+  
+    // Example: 0x2::coin::Coin<0x2::sui::SUI>
+    const genericMatch = typeString.match(/<(.+)>/);
+    if (genericMatch) {
+      const genericType = genericMatch[1];
+      console.log('Generic type parameter:', genericType);
+      return genericType;
+    }
```

Comment:
Can you extract this into a fn, or the regex to understand better what's happening? Does Sui library doesn't have any internal fn to handle this ?